src/rpc_client/src/meta_client.rs

```diff
@@ -851,93 +868,335 @@ impl HummockMetaClient for MetaClient {
+                let mut ddl_c = self.ddl_client.as_ref().lock().await;
+                let mut hummock_c = self.hummock_client.as_ref().lock().await;
+                let mut notification_c = self.notification_client.as_ref().lock().await;
+                let mut stream_c = self.stream_client.as_ref().lock().await;
+                let mut user_c = self.user_client.as_ref().lock().await;
+                let mut scale_c = self.scale_client.as_ref().lock().await;
+                let mut backup_c = self.backup_client.as_ref().lock().await;
+
+                *leader_c = LeaderServiceClient::new(leader_channel.clone());
+                *cluster_c = ClusterServiceClient::new(leader_channel.clone());
+                *heartbeat_c = HeartbeatServiceClient::new(leader_channel.clone());
+                *ddl_c = DdlServiceClient::new(leader_channel.clone());
+                *hummock_c = HummockManagerServiceClient::new(leader_channel.clone());
+                *notification_c = NotificationServiceClient::new(leader_channel.clone());
+                *stream_c = StreamManagerServiceClient::new(leader_channel.clone());
+                *user_c = UserServiceClient::new(leader_channel.clone());
+                *scale_c = ScaleServiceClient::new(leader_channel.clone());
+                *backup_c = BackupServiceClient::new(leader_channel);
+            } // release MutexGuards on all clients
+
+            tracing::info!("Successfully failed over meta_client to {}", addr);
+            return;
+        }
+
+        // Only print failure messages if the entire failover failed
+        failure_msgs.sort_unstable();
+        failure_msgs.dedup();
+        tracing::error!("Failed to failover to new meta leader. Error messages were...");
+        for msg in failure_msgs {
+            tracing::error!(msg);
+        }
+    }
+
+    /// Execute the failover if it is needed. If no failover is needed do nothing
+    /// Returns true if failover was needed, else false
+    pub async fn do_failover_if_needed(&self) -> bool {
+        let current_leader = self.try_get_leader_from_connected_node().await;
+        let current_leader_clone = current_leader.clone();
+        let connected_node = self.leader_address.as_ref().lock().await.clone();
+        if current_leader.is_some() && current_leader.unwrap() == connected_node {
+            return false;
+        }
+        tracing::warn!(
+            "failing over meta_client. Old meta node was {:?}",
+            connected_node,
+        );
+        self.do_failover(current_leader_clone).await;
+        true
+    }
+
+    /// Connect to the meta server `addr`.
+    pub async fn new(addr: &str) -> Result<Self> {
+        tracing::info!("GrpcMetaClient originally connected against {}", addr);
+        let channel = get_channel_with_defaults(addr).await?;
+
+        // Dummy address forces a client failover
+        let dummy_address = HostAddress {
```

Comment:
Why don't get leader address directly here? Storing a dummy info here is quite weird.